drivers: stepper: adi: trinamic tmc5041

jilaypandya opened the pull request:
This PR does the following:
1. Introduce initial structure for trinamic stepper drivers.
2. TMC5041 driver.

Todos:

- [x] Add this driver to build once the file structutre and API changes have been agreed upon
- [x] Add invert direction to stepper-controller.yaml

jilaypandya committed:
drivers: stepper: adi: trinamic tmc5041

This commit introduces initial structure for trinamic drivers
TMC5041 is implemented with following features:
- StallGuard
- RAMPSTAT_POLL
- RAMP_GEN

Signed-off-by: Dipak Shetty <dipak.shetty@zeiss.com>
Signed-off-by: Jilay Pandya <jilay.pandya@zeiss.com>

bjarki-andreasen reviewed (dismissed):
Device driver specific APIs go to include/zephyr/drivers/stepper/ and are not exposed using an API structure, they are just "normal" functions :)

jilaypandya committed:
drivers: stepper: test: build_all: tmc5041

This commit add tmc5041 to stepper build all tests

Signed-off-by: Jilay Pandya <jilay.pandya@zeiss.com>

bjarki-andreasen commented on drivers/stepper/adi/Kconfig:
```diff
@@ -0,0 +1,50 @@
+	help
+	  Enable trinamic stepper controller
+
+config TRINAMIC_RAMP_GENERATOR
```
maybe use ADI_TMC as well to match the file structure `adi/tmc_...` ?

bjarki-andreasen commented on dts/bindings/stepper/adi/adi,trinamic-tmc5041.yaml:
```diff
@@ -0,0 +1,79 @@
+
+description: Trinamic TMC5041 Stepper Motor Controller
+
+compatible: "adi,trinamic-tmc5041"
```
This should be adi,tmc5041 which is vendor,model

bjarki-andreasen commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,73 @@
+ */
+#define TMC_RAMP_DT_SPEC_GET(node)                                                                 \
+	{                                                                                          \
+		.tzerowait = DT_PROP_OR(node, tzerowait, 0),                                       \
+		.vstart = DT_PROP_OR(node, vstart, 0), .a1 = DT_PROP_OR(node, a1, 0),              \
+		.amax = DT_PROP_OR(node, amax, 0), .d1 = DT_PROP_OR(node, d1, 0),                  \
+		.dmax = DT_PROP_OR(node, dmax, 0), .v1 = DT_PROP_OR(node, v1, 0),                  \
+		.vmax = DT_PROP_OR(node, vmax, 0), .vstop = DT_PROP_OR(node, vstop, 0),            \
+		.vhigh = DT_PROP_OR(node, vhigh, 0), .vcoolthrs = DT_PROP_OR(node, vcoolthrs, 0),  \
+		.iholdrun = (TMC5041_IRUN(DT_PROP_OR(node, irun, 0)) |                             \
+			     TMC5041_IHOLD(DT_PROP_OR(node, ihold, 0)) |                           \
+			     TMC5041_IHOLDDELAY(DT_PROP_OR(node, iholddelay, 0))),                 \
```
Could you format this so it matches the layout of `struct tmc_ramp_generator_data` ?

```
	{                 
		.vstart = DT_PROP_OR(node, vstart, 0),
                .a1 = DT_PROP_OR(node, a1, 0),
                ... 
        }
```
that makes it easy to read/review since you can go one by one without searching :)

bjarki-andreasen commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,73 @@
+			     TMC5041_IHOLDDELAY(DT_PROP_OR(node, iholddelay, 0))),                 \
+	}
+
+#ifdef CONFIG_DT_HAS_ADI_TMC5041_ENABLED
```
No need to ifdef, the linker will tell you that the function is missing, which is a clear error, compared to whatever the compiler will produce from the undefined symbol tmc5041_stepper_set_ramp

faxe1008 commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
+	(void)tmc5041_read(config->controller, TMC5041_SWMODE(config->index), &reg_value);
```
This needs to check the error, in the fringe case that fail but writes suceed, this wipes all other flags.

faxe1008 commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct tmc5041_stepper_data *data =
+		CONTAINER_OF(dwork, struct tmc5041_stepper_data, stallguard_dwork);
+	LOG_DBG("Stallguard work handler");
```
I'd inline print_stall_status so you get this sort of for free since the method name is prepended.

faxe1008 commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+
+		if (ramp_stat_interrupt_values > 0) {
+			switch (ramp_stat_interrupt_values) {
+			case BIT(0):
```
Defines with good names would be more readable here imo.

faxe1008 commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+			return -EINVAL;
+		}
+		LOG_DBG("INT pin callback set");
+		uint32_t read_value;
```
I guess this is flagged by checkpatch?

faxe1008 commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+{
+	LOG_DBG("Stepper motor controller %s %s", dev->name, enable ? "enabled" : "disabled");
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
```
newline missing.

fabiobaltieri replied:
drop the init, more of these, you got the idea

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int status = k_mutex_lock(&data->mutex, K_MSEC(100));
```
Do you really want the timeout? you are ignoring the return value on all callers that I can tell and there's no warning here so this has no chance of timing out and being discovered

How about using `K_FOREVER` here, adding a `LOG_WRN` to the error handling and change the return to void and ditch all those `(void)` casting? Or change the callers to get the error code and log something.

Would clean the code below as well, normally you want to branch off error handling and keep the normal execution inline.

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int status = k_mutex_lock(&data->mutex, K_MSEC(100));
```
same same

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+#else
+		ARG_UNUSED(data);
+		LOG_WRN("STEPPER_ADI_TMC5041_INT not enabled, async signal not supported");
+#endif
```
return -1? and log the warning on the caller I'd say, or change to void and leave the warning here

the whole return code handling here is a bit unclear, either check them or get rid of them, your choice

jilaypandya replied:
changing to void with a LOG_WARN

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+
+	int32_t sg_result = 0;
```
no need to init this, or rather it should not be needed, right now there's some situation where an the read fails and this returns zero with no logging or anything

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+static int stallguard_enable(const struct device *dev, const bool enable)
+{
+	const struct tmc5041_stepper_config *config = dev->config;
+	uint32_t reg_value = 0;
```
drop the init

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+	struct tmc5041_data *data = CONTAINER_OF(cb, struct tmc5041_data, int_cb);
+	(void)k_work_submit(&(data->int_pin_callback_work));
```
just `&data->int_pin_callback_work`

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+		uint32_t drv_status;
+
+		(void)tmc5041_read(data->controller, TMC5041_DRVSTATUS(idx), &drv_status);
+		if (1U == FIELD_GET(TMC5041_DRVSTATUS_SG_MASK, drv_status)) {
```
eeh, I think I know why but that's a bit uncommon in the code base :-) how about a more traditional

```c
if (FIELD_GET(TMC5041_DRVSTATUS_SG_MASK, drv_status) == 1U) {
```

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+		k_work_init(&controller_data->int_pin_callback_work, callback_work_handler);
+	}
+
+	return status;
```
`return 0`, drop the init

fabiobaltieri commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+
+#define TMC5041_STEPPER_DATA_DEFINE(child)							\
+	static struct tmc5041_stepper_data tmc5041_stepper_data_##child = {			\
+		IF_ENABLED(CONFIG_STEPPER_ADI_TMC_STALLGUARD,					\
+			(.stepper = DEVICE_DT_GET(child),))};
```
shouldn't this be in the config structure?

jilaypandya replied:
i need access to device* in k_work_handler(_stallguard_handler_) to enable/disable stallguard for that particular stepper motor

bjarki-andreasen replied:
usually this is done by having a pointer to the parent device in the data struct, and then getting the config through it :)

```
struct foo_dev_config {
        ...
};

struct foo_dev_data {
        const struct device *dev;
}

void foo(struct foo_dev_data *data)
{
        const struct device dev = data->dev;
        const struct foo_dev_config config = dev.>config;
        ...
}
```

jilaypandya replied:
I am doing the same as well, i need to access the the config of that device.
```
#define TMC5041_STEPPER_DATA_DEFINE(child)							\
	static struct tmc5041_stepper_data tmc5041_stepper_data_##child = {			\
		.stepper = DEVICE_DT_GET(child),};

static void stallguard_work_handler(struct k_work *work)
{
	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
	struct tmc5041_stepper_data *stepper_data =
		CONTAINER_OF(dwork, struct tmc5041_stepper_data, stallguard_dwork);
	int err;
	const struct tmc5041_stepper_config *stepper_config = stepper_data->stepper->config;

	LOG_DBG("Stallguard work handler");
#if CONFIG_STEPPER_LOG_LEVEL >= LOG_LEVEL_DBG
	int32_t sg_result;

	err = tmc5041_read(stepper_config->controller,
			   TMC5041_DRVSTATUS(stepper_config->index), &sg_result);
	if (err) {
		LOG_ERR("Failed to read DRVSTATUS register");
		return;
	}
	LOG_DBG("Detecting Load for motor %s %ld %ld", stepper_data->stepper->name,
		sg_result & TMC5041_DRV_STATUS_SG_RESULT_MASK,
		(sg_result & TMC5041_DRV_STATUS_SG_STATUS_MASK) >>
			TMC5041_DRV_STATUS_SG_STATUS_SHIFT);
#endif

	err = stallguard_enable(stepper_data->stepper, true);
	if (err == -EAGAIN) {
		LOG_ERR("Trying again to enable stallguard");

		k_work_reschedule(dwork, K_MSEC(stepper_config->stallguard_delay_in_ms));
	}
	if (err == -EIO) {
		LOG_ERR("Failed to enable stallguard becaoause of I/O error");
		return;
	}
}

```

bjarki-andreasen replied:
The struct tmc5041_stepper_data is not const, but .stepper is, you are storing immutable data in a mutable data structure (RAM)

fabiobaltieri commented on drivers/stepper/adi/tmc_spi.c:
```diff
@@ -0,0 +1,119 @@
+			  uint32_t *data)
+{
+	uint8_t tx_buffer[BUFFER_SIZE] = {0x7FU & register_address, 0U, 0U, 0U, 0U};
+	uint8_t rx_buffer[BUFFER_SIZE] = {0U};
```
no need to initialize this one

fabiobaltieri commented on drivers/stepper/adi/tmc_spi.c:
```diff
@@ -0,0 +1,119 @@
+int tmc_spi_read_register(const struct spi_dt_spec *bus, const uint8_t register_address,
+			  uint32_t *data)
+{
+	uint8_t tx_buffer[BUFFER_SIZE] = {0x7FU & register_address, 0U, 0U, 0U, 0U};
```
add a define for that `0x7fu`

fabiobaltieri commented on drivers/stepper/adi/tmc_spi.h:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @return a value from spi_transceive().
+ */
+int tmc_spi_read_register(const struct spi_dt_spec *bus, const uint8_t register_address,
```
is this good for all TMC devices? wondering if it should be `tmc5xxx` instead

dipakgmx replied:
Think, it should work. Ran a check for the `ADDRESS_MASK` on the [Trinamic IC folder](https://github.com/analogdevicesinc/TMC-API/tree/master/tmc/ic): 

```
grep -rInH --include="*.h" ".ADDRESS_MASK" ic
ic/TMC2160/TMC2160_HW_Abstraction.h:18:#define TMC2160_ADDRESS_MASK     0x7F
ic/TMC2300/TMC2300_HW_Abstraction.h:17:#define TMC2300_ADDRESS_MASK     0x7F
ic/TMC5262/TMC5262_HW_Abstraction.h:18:#define TMC5262_ADDRESS_MASK     0x7F
ic/TMC2226/TMC2226_HW_Abstraction.h:18:#define TMC2226_ADDRESS_MASK     			0x7F
ic/TMC7300/TMC7300_HW_Abstraction.h:16:#define TMC7300_ADDRESS_MASK     0x7F
ic/TMC2660/TMC2660_Macros.h:16:#define TMC2660_ADDRESS(datagram) (((datagram) >> TMC2660_ADDRESS_SHIFT) & TMC2660_ADDRESS_MASK)
ic/TMC2660/TMC2660_Constants.h:18:#define TMC2660_ADDRESS_MASK     0x07
ic/TMC5271/TMC5271_HW_Abstraction.h:18:#define TMC5271_ADDRESS_MASK     0x7F
ic/TMC5240/TMC5240_HW_Abstraction.h:17:#define TMC_ADDRESS_MASK         0x7F
ic/TMC5072/TMC5072_HW_Abstraction.h:18:#define TMC5072_ADDRESS_MASK     0x7F
ic/TMC5130/TMC5130_HW_Abstraction.h:16:#define TMC5130_ADDRESS_MASK     0x7F
ic/TMC4671/TMC4671_Fields.h:25:#define TMC4671_CHIP_INFO_ADDRESS_MASK                       0xFF // CHIPINFO_ADDR //
ic/TMC6100/TMC6100_HW_Abstraction.h:16:#define TMC6100_ADDRESS_MASK     0x7F
ic/TMC2130/TMC2130_HW_Abstraction.h:18:#define TMC2130_ADDRESS_MASK     0x7F
ic/TMC4361A/TMC4361A_Constants.h:18:#define TMC4361A_ADDRESS_MASK     TMC_ADDRESS_MASK
ic/TMC2590/TMC2590_Macros.h:18:#define TMC2590_GET_ADDRESS(X) ((uint8_t) ((((X) >> (TMC2590_ADDRESS_SHIFT + 1)) ? ((X) >> TMC2590_ADDRESS_SHIFT) : 0) & TMC2590_ADDRESS_MASK))
ic/TMC2590/TMC2590_Constants.h:18:#define TMC2590_ADDRESS_MASK     0x07
ic/TMC5031/TMC5031_HW_Abstraction.h:17:#define TMC5031_ADDRESS_MASK     0x7F
ic/TMC2225/TMC2225_HW_Abstraction.h:18:#define TMC2225_ADDRESS_MASK     0x7F
ic/TMC5062/TMC5062_HW_Abstraction.h:18:#define TMC5062_ADDRESS_MASK     0x7F
ic/TMC2240/TMC2240_HW_Abstraction.h:17:#define TMC2240_ADDRESS_MASK     0x7F
ic/TMC2224/TMC2224_HW_Abstraction.h:19:#define TMC2224_ADDRESS_MASK     0x7F
ic/TMC5041/TMC5041_HW_Abstraction.h:18:#define TMC5041_ADDRESS_MASK     0x7F
ic/TMC2208/TMC2208_HW_Abstraction.h:18:#define TMC2208_ADDRESS_MASK     0x7F
ic/TMC6200/TMC6200_HW_Abstraction.h:16:#define TMC6200_ADDRESS_MASK     0x7F
ic/TMC2262/TMC2262_HW_Abstraction.h:15:#define TMC2262_ADDRESS_MASK     0x7F
ic/TMC5272/TMC5272_HW_Abstraction.h:18:#define TMC_ADDRESS_MASK         0x7F
ic/TMC2209/TMC2209_HW_Abstraction.h:17:#define TMC_ADDRESS_MASK         0x7F
ic/TMC5160/TMC5160_HW_Abstraction.h:17:#define TMC5160_ADDRESS_MASK     0x7F
```

fabiobaltieri replied:
brilliant, still I would rename this to `adi_tmc`

trinamic-ed replied:
yes, the "Trinamic Motion Control" team has been integrated into ADI as "ADI Trinamic", thus "adi_tmc" makes sense. For some new ICs like TMC8100 and TMC9660 the ADDRESS_MASK will be different.

jilaypandya replied:
Ahh okay, how do we handle this correctly? I'll extend the read write functions to include address masks and then every driver can handle this locally

jilaypandya commented on drivers/stepper/adi/Kconfig:
```diff
@@ -0,0 +1,50 @@
+	help
+	  Enable trinamic stepper controller
+
+config STEPPER_ADI_TMC_RAMP_GEN
+	bool "Use Trinamic Stepper Controller with Ramp Generator"
+	depends on STEPPER_ADI_TMC
+	default y
+	help
+	  If y, the Trinamic Stepper Controller uses Ramp Generator
+
+config STEPPER_ADI_TMC_STALLGUARD
+	bool "Use Trinamic Stepper Controller with StallGuard"
+	depends on STEPPER_ADI_TMC
+	default y
+	help
+	  If y, the Trinamic Stepper Controller uses StallGuard
+
+if STEPPER_ADI_TMC_STALLGUARD
+
+config STEPPER_ADI_TMC_STALLGUARD_ACTIVATION_PAUSE_IN_MSEC
+	int "StallGuard Activation Pause in ms"
+	depends on STEPPER_ADI_TMC_STALLGUARD
+	default 100
+	help
+	  StallGuard Activation Pause in ms.
+	  StallGuard should not be enabled during motor spin-up.
+	  Wait for the motor to reach a stable speed before enabling StallGuard.
+
+endif # STEPPER_ADI_TMC_STALLGUARD
```
would it be a good idea to use 
`DT_STRING_UNQUOTED(node_id, prop)
`
then in device tree you could use something like 
```
stepper: stepper@0{
     use-stallguard;
     stallguard-delay = "K_MSEC(100)";
}
```

gmarull replied:
I was not even sure what to say when I say `stallguard-delay = "K_MSEC(100)";` on my inbox. Please keep in mind DT needs to avoid any sort of zephyrisms, hacks, etc. Just use `stallguard-delay-ms = <100>`.

jilaypandya replied:
The same could be applied for the RAMP_GENERATOR part as well, where you could activate ramp-gen from the dt bindings. This entire thing could be removed from kconfig alltogether then i suppose

gmarull replied:
The "same" can't be applied to any property. Please read the DT-spec or bindings docs. `"K_MSEC(100)";` in DT just doesn't make any sense.

bjarki-andreasen replied:
To summarize a bit, things that describe hardware, can be in the devicetree. Things like interrupt routing, a flag indicating whether the specific hardware supports some feature, how many of some thing that are present in a specific instance of hardware, like number of RX mailboxes of a specific CAN instance.

Things that do not go in devicetree are anything that is specific to a driver or subsystem, like the size of some buffer, or lazy delay used in a driver.

There are exceptions to everything, but generally, anything in the devicetree should be usable by any driver, even a non-zephyr driver.

fabiobaltieri replied:
> would it be a good idea to use
> `DT_STRING_UNQUOTED(node_id, prop) `

Yeah, no don't do that :-)

`stallguard-delay-ms = <xxx>;`

jilaypandya commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	return status;
+}
+
+static void calculate_velocity_from_hz_to_fclk(const struct device *dev, const uint32_t velocity_hz,
```
Why? I mean if the user want to input zero, its okay, there wont be a zero division error, or am i overseeing something

bjarki-andreasen replied:
Use a build assert to ensure the clock is not zero and remove the check, see my comment  a few lines down :)

jilaypandya commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+static int32_t print_stall_status(const struct device *dev)
```
changing to void

jilaypandya replied:
function removed since it's only used/required only in one particular debug mode when the stallguard is enabled.

dipakgmx commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+		steps);
+	(void)tmc5041_write(config->controller, TMC5041_XTARGET(config->index), target_position);
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)k_work_reschedule(&data->stallguard_dwork,
```
The stall guard activation delay is more speed dependent.
Am not sure if having a fixed delay is a good idea here.

jilaypandya replied:
yep you are correct, i am introducing 
`.stallguard_threshold_velocity = DT_PROP_OR(child, stallguard_threshold_velocity, 0),`

jilaypandya replied:
so the stallguard work handler would retry activation again and again like this. Normally in the drive stall should be detected and appropriately calibrated, if that is not happening then the device is anyhow not correctly calibrated.

	"err = stallguard_enable(data->stepper, true);
	if (err == -EAGAIN) {
		LOG_ERR("Trying again to enable stallguard");
		struct tmc5041_stepper_config *stepper_config = data->stepper->config;
		k_work_reschedule(dwork, K_MSEC(stepper_config->stallguard_delay_in_ms));
	}
	if( err == -EIO) {
		LOG_ERR("Failed to enable stallguard becaoause of I/O error");
		return;
	}"

dipakgmx commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	return 0;
+}
+
+static int tmc5041_stepper_set_max_velocity(const struct device *dev, const uint32_t velocity)
```
`const uint32_t velocity`doesn't match with the [API](https://docs.zephyrproject.org/latest/doxygen/html/group__stepper__interface.html#gabbb691c2f1beb2bdc7e856a7f77b4de4).

dipakgmx commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+	return 0;
+}
+
+static int tmc5041_stepper_set_micro_step_res(const struct device *dev,
```
[API](https://docs.zephyrproject.org/latest/doxygen/html/group__stepper__interface.html#ga6645255ed9249f963273e13f7bd09ff8) doesnt match here.

dipakgmx commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+
+	LOG_DBG("Controller: %s, Stepper: %s", stepper_config->controller->name, dev->name);
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)tmc5041_write(stepper_config->controller, TMC5041_SWMODE(stepper_config->index), 0);
```
Why activate stallguard, even before a move operation?

jilaypandya replied:
I would activate stallguard before any kind of move operation. or?

dipakgmx replied:
yes

bjarki-andreasen commented on drivers/stepper/adi/tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,608 @@
+{
+	const struct tmc5041_config *config = dev->config;
+
+	if (config->clock_frequency == 0) {
+		LOG_ERR("Clock frequency cannot be zero");
+		return;
+	}
```
This should be a build assert, no need to check this const value at runtime :) just add `BUILD_ASSERT(DT_INST_PROP(clock_frequency), "clock frequency must be positive");` to the init macro

jilaypandya commented:
**Update:**
I have removed the interrupt detection for now. Will introduce that feature in an upcoming PR most probably. However there is a provision to poll the RAMP_STAT register and check for pos_reached or stall event

dipakgmx commented on drivers/stepper/adi_tmc/ad_tmc5041_stepper_controller.c:
File name seems incorrect. Needs to be `adi_tmc5041_stepper_controller.c`.

jilaypandya replied:
i took the reference from sensors actually. over there they have names such adltc2990.c and so on, in the folder adi they use prefix ad for analog devices. Probably let's keep it that way for now?, or?

jilaypandya replied:
On a second thought, adi_tmc sounds better

dipakgmx replied:
is the `ad_` prefix needed in the first place?

jilaypandya replied:
https://github.com/zephyrproject-rtos/zephyr/pull/79163#discussion_r1781554692

dipakgmx commented on drivers/stepper/adi_tmc/ad_tmc_reg.h:
File needs to to have tmc5041 in its name.

jilaypandya replied:
I was thinking of putting all the registers in a single file like this, I think a lot of these registers are repetitive and could be used across drivers.
```
#ifdef CONFIG_STEPPER_ADI_TMC5041

#define TMC5041_MOTOR_ADDR(m)     (0x20 << (m))
#define TMC5041_MOTOR_ADDR_DRV(m) ((m) << 4)
#define TMC5041_MOTOR_ADDR_PWM(m) ((m) << 3)
```

faxe1008 commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,747 @@
+		return -EIO;
+	}
+
+	*is_moving = (FIELD_GET(TMC5041_DRV_STATUS_STST_SHIFT_BIT, reg_value) == 1U) ? true : false;
```
Ternary not needed here the result is already bool

faxe1008 commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,747 @@
+	const struct tmc5041_config *config = dev->config;
+	int err;
+
+	k_sem_init(&data->sem, 0, 1);
```
You could add a static declaration of the semaphore to the device definition macro and only store a pointer to it in the data, this would remove the need for this initialisation. Or just combine the two lines by setting the initial count to 1 straight away.

dipakgmx commented on dts/bindings/stepper/adi/adi,trinamic-gconf.yaml:
```diff
@@ -0,0 +1,89 @@
+  shaft2:
+    type: boolean
+    description: |
+      1: Inverse motor 2 direction
```
The shaft direction inversion properties are better suited at [stepper-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/stepper/stepper-controller.yaml)

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+
+struct tmc_ramp_generator_data {
+	uint32_t vstart;
+	uint32_t a1;
```
uint16_t

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+	uint32_t a1;
+	uint32_t v1;
+	uint32_t d1;
+	uint32_t amax;
```
uint16_t

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+	uint32_t d1;
+	uint32_t amax;
+	uint32_t vmax;
+	uint32_t dmax;
```
uint16_t

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+	uint32_t vmax;
+	uint32_t dmax;
+	uint32_t vstop;
+	uint32_t tzerowait;
```
uint16_t

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+	uint32_t vstart;
+	uint32_t a1;
+	uint32_t v1;
+	uint32_t d1;
```
uint16_t

dipakgmx commented on include/zephyr/drivers/stepper/stepper_trinamic.h:
```diff
@@ -0,0 +1,74 @@
+		.vhigh = DT_PROP_OR(node, vhigh, 0),					\
+		.iholdrun = (TMC5041_IRUN(DT_PROP_OR(node, irun, 0)) |			\
+			     TMC5041_IHOLD(DT_PROP_OR(node, ihold, 0)) |		\
+			     TMC5041_IHOLDDELAY(DT_PROP_OR(node, iholddelay, 0))),	\
```
Need to add assertions that values provided are within range

dipakgmx replied:
Assertions need to be included to ensure that the parameters A1, D1, V1, etc are within the limits.

bjarki-andreasen commented on dts/bindings/stepper/stepper-controller.yaml:
```diff
@@ -6,6 +6,11 @@ description: Stepper Controller
 properties:
+  invert-direction:
+    type: boolean
+    description: |
+      1: Inverse motor direction
```
`1:` is redundant :)

jilaypandya committed:
shell: stepper: add further stepper signals to shell

This commit adds further signals to stepper shell
- STEPPER_SIGNAL_SENSORLESS_STALL_DETECTED
- STEPPER_SIGNAL_LEFT_END_STOP_DETECTED
- STEPPER_SIGNAL_RIGHT_END_STOP_DETECTED

Signed-off-by: Jilay Pandya <jilay.pandya@zeiss.com>

jilaypandya commented:
**Update:** 
- Have adjusted the shell to read the newly added stepper async signals

Won't be adding any more features to this PR :-)

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+	err = tmc_spi_write_register(&bus, TMC5041_WRITE_BIT, reg_addr, reg_val);
+	if (err) {
+		LOG_ERR("Failed to write register 0x%x with value 0x%x", reg_addr, reg_val);
+	}
+	k_sem_give(&data->sem);
```
```c
	}

	k_sem_give(&data->sem);
```

nit: blank line for symmetry, same below

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+{
+	struct tmc5041_stepper_data *data = dev->data;
+
+	if (async) {
```
swap the condition, keep the normal code path unindented (well, less indented)

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+	}
+	k_sem_give(&data->sem);
+
+	return err;
```
`return 0` (wherever you can in the file, that way it's unambiguous that the function termination is always the success case and the errors are branched out)

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+	LOG_DBG("Controller: %s, Stepper: %s", stepper_config->controller->name, dev->name);
+
+	if (stepper_config->is_sg_enabled) {
+
```
stray blank line, drop

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+}
+
+#define TMC5041_SHAFT_CONFIG(child)								\
+	(DT_PROP(child, invert_direction) << TMC5041_GCONF_SHAFT_SHIFT(DT_REG_ADDR(child)))|
```
add a space before the `|`

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c:
```diff
@@ -0,0 +1,764 @@
+	IF_ENABLED(CONFIG_STEPPER_ADI_TMC_RAMP_GEN, (CHECK_RAMP_DT_DATA(child)));		\
+	static const struct tmc5041_stepper_config tmc5041_stepper_config_##child = {		\
+		.controller = DEVICE_DT_GET(DT_PARENT(child)),					\
+		.default_micro_step_res = DT_PROP_OR(child, micro_step_res,			\
+						     STEPPER_MICRO_STEP_256),			\
```
I would lay on the side of putting the default in the devicetree rather than using `DT_PROP_OR`, better location for defaults than the driver code, same for the other props, makes this macro a bit more compact if anything

jilaypandya replied:
I was of the understanding that the defaults in the device tree are the register values after power-on. If this is allowed, i would be more than glad to put the defaults in the bindings.

jilaypandya replied:
Have added defaults in bindings

fabiobaltieri reviewed (commented):
cool, few more comments, mostly styling

jilaypandya commented on drivers/stepper/adi_tmc/Kconfig:
```diff
@@ -0,0 +1,52 @@
+	help
+	  Enable trinamic stepper controller
+
+config STEPPER_ADI_TMC_RAMP_GEN
```
yeah that's correct, the entire menu is then shifted to a new page, looks much better
